src/ipips/ipip-0431.md

```diff
@@ -0,0 +1,197 @@
+
+The proposed solution introduces a new parameter for the CAR content type in HTTP requests
+and responses: `meta`.
+
+The `meta` parameter allows clients to request the server to include additional metadata about the CAR along with the response body.
+
+The value of this parameter includes both the location where the metadata is given (e.g. `eof`) as well as the type of data received (e.g. `json`) separated by a `+`, to give a value such as `meta=eof+json`
+
+When the location parameter is set to `eof`, which is currently the only supported value, the server SHOULD respond with the following response body:
+
+```
+<Response body as CARv1 stream> <0x00 byte> <Metadata>
+```
+
+The only supported value for the data type parameter is `json`. This signifies that the metadata MUST be a JSON object.
+
+This parameter MUST only be used with CAR `version=1`.
+
+When the parameter is not set or does not equal `eof+json`, the server SHOULD not add any extra blocks to the response, neither the 0x00 byte nor any metadata.
+
+This results in a example content type of `application/vnd.ipld.car;version=1;meta=eof+json`
+
+See [CAR `meta` (content type parameter)](/http-gateways/trustless-gateway/#car-meta-content-type-parameter)
+in Trustless Gateway specification for more details.
+
+## Design rationale
+
+The proposal introduces a minimal change allowing Gateways and retrieval clients to explicitly opt
+into receiving additional metadata block at the end of the CAR response stream.
+
+The metadata block is designed to be very flexible and able to support new use-cases that may arise
+in the future.
+
+### User benefit
+
+- Clients of trustless gateways can use the fields from the metadata as an attestation that they
+performed the retrieval from the given server.
+
+- For example, the metadata block could include a `car_bytes` field, the byte length of the CAR stream (excluding the metadata block). This would allow clients to verify whether they received all CAR
+bytes, which provides a backward-compatible solution for the [CARv1 streaming problem](https://github.com/ipfs/specs/pull/332) until new CAR version is introduced.
+
+- As another example, the metadata object includes the `error` field, allowing the server to pass back additional information about why the response is an error, such as why the CAR stream was incomplete.
+
+- In the SPARK use case, retrieval clients would like to prove they have retrieved an entire file from a specific retrieval provider that has implemented the trustless gateway spec. The additional metadata block allows checksums and signatures to be passed along with the data, allowing the retrieval client to create a proof of correct retrieval.
+
+- The metadata `sig` field SHOULD also be populated, returning a signature, using the server's Ed2559 identity, over the metadata properties object. This allows gateway clients to submit the metadata block as an attestation of retrieval that 3rd parties can verify.
+
+### Compatibility
```

Comment:
@bajtos  Let's go extra mile here and elaborate what happens when CAR response with  `0x00`-prefixed suffix  is parsed by existing CAR software.

My suggestion is to add some clear statement about expected interop, like "libraries and implementations SHOULD ignore the suffix after 0x00", otherwise we will create a bad UX/DX, where developer tries to debug things with existign tooling and the tooling errors.

Personally, I imagine we don't want things to fail due to `0x00` suffix, namely:

- \>80% of Amino DHT IPFS network (including IPFS Desktop and Brave) is  Kubo
  - `ipfs dag import` should ignore suffix
- reference CAR libraries 
  - [js-car](https://github.com/ipld/js-car) (JS library used by things like custom Service Workers, Helia)
  - [go-car](https://github.com/ipld/go-car)  v1 and v2  (GO libraries)
- CLI tools we recommend to developers, they will try to use these for debugging CAR responses with the suffix:
  - [ipfs-car](https://www.npmjs.com/package/ipfs-car) (JS CLI) 
  - [car](https://github.com/ipld/go-car/tree/master/cmd/car#readme)  (CLI) 
  - [go-fixtureplate](https://github.com/ipld/go-fixtureplate) (CLI)

Reply:
> Let's go extra mile here and elaborate what happens when CAR response with 0x00-prefixed suffix is parsed by existing CAR software.

It's a great idea to think about compatibility with existing & future tooling and clearly describe our thinking. 👍🏻 

The most important aspect is avoiding the "0x00 insertion attack" vector. You can find more details in the section [Zero-length-block insertion attacks](https://github.com/ipfs/specs/blob/b6069bf0d0d9ff57ebdba36cf57f0d2f079bef1f/src/ipips/ipip-0431.md#zero-length-block-insertion-attacks) (including the Filecoin-specific logic). I am cross-posting the mitigation I proposed:

>  Our proposal avoids this attack vector:
>  - It does not change the current semantics of CARv1. Zero-length blocks remain invalid.
>  - Instead, we treat the response body as a new container format combining the CARv1 file with additional data.
>  - Clients must explicitly request this new container format. Existing clients not aware of the new metadata will not receive responses in the new format.

When developers use existing tooling, they will never receive a CAR file with the `0x00` suffix.

There are two major ways how a CAR with a `0x00` suffix can emerge:

1. Somebody makes an HTTP request to a Trustless Gateway, explicitly asks to receive CAR with `meta=eof+json`, saves the response body to a `.car` file and forgets to extract the CAR payload from the container (remove the `\x00{metadata}` trailer).

2. Somebody uses a tool that is aware of `meta=eof+json`. The tool opts into this new feature when requesting content from a Trustless Gateway, but does not extract the CAR payload from the container in the response body before returning the content back to the user.

I am arguing that (2) is a bug in the tooling, introduced by the change that modified Trustless Gateway requests to opt-into `meta=eof+json`, and therefore, the maintainers of that tool should fix that bug - make the tool adhere to spec.

Regarding (1): do you think this will happen frequently enough to justify the effort required to change all libraries you mentioned to start ignoring the `0x00` byte?

Maybe it's actually a good thing that the tooling reports an error because it tells the user they are using the new `meta=eof+json` feature incorrectly.

As an alternative to silently stripping the `0x00` suffix, the tooling can detect the situation where `0x00` is followed by a valid DAG-JSON object and report a more helpful error message to the user, advising them to either change the "accept" header in the request to the Trustless Gateway or else remove the `0x00` suffix (unpack CARv1 from the container format).

Thoughts?

---

`go-car/cmd/car/inspect.go` seems to always treat `0x00` as EOF, if I am reading the source code correctly:

https://github.com/ipld/go-car/blob/5c5d432d582564f88fd2124f2fce4f2f3e47a654/cmd/car/inspect.go#L26
```go
	rd, err := carv2.NewReader(inStream, carv2.ZeroLengthSectionAsEOF(true))
```

`js-car` seems to always reject zero-length blocks:

https://github.com/ipld/js-car/blob/562c39266edda8422e471b7f83eadc8b7362ea0c/src/decoder.js#L94-L97
```js
  let length = decodeVarint(await reader.upTo(8), reader)
  if (length === 0) {
    throw new Error('Invalid CAR section (zero length)')
  }
```

I guess I can test how existing tooling handles zero-length blocks and document this behaviour in the IPIP, so that we better understand the current landscape.